Refactor Draggable to decouple drag handle from the DOM node being dragged

oandregal opened the pull request:
Related https://github.com/WordPress/gutenberg/issues/7114
Makes https://github.com/WordPress/gutenberg/pull/8764 obsolete

This PR refactors the internals of drag-and-drop, to make easier changing the drag handle in next iterations. Note that the external behavior of the drag-and-drop event is not changed in this PR.

### The `Draggable` component

It refactors the `Draggable` component to decouple the drag handle (the wrapped component) from the element being dragged. After this change, `Draggable` is no longer be concerned with creating an actual DOM node that serves as drag handle, but the drag handle is the wrapped component itself. This makes `Draggable` reusable in use cases other than making a block draggable.

The existing behavior is deprecated and planned to be removed on the next version. 

How to use it:

```jsx
import { Dashicon, Draggable, Panel, PanelBody } from '@wordpress/components';

const MyDraggable = ( { onDragStart, onDragEnd } ) => (
	<div id="draggable-panel">
		<Panel header="Draggable panel" >
			<PanelBody>
				<Draggable
					elementId="draggable-panel"
					transferData={ { } }
					onDragStart={ onDragStart }
					onDragEnd={ onDragEnd }
				>
				{
					( { onDraggableStart, onDraggableEnd } ) => (
						<Dashicon
							icon="move"
							onDragStart={ onDraggableStart }
							onDragEnd={ onDraggableEnd }
							draggable
							/>
					)
				}
				</Draggable>
			</PanelBody>
		</Panel>
	</div>
);

export default MyDraggable;
```

### Test

Drag and drop a block and test that it works as used to be (the block being dragged is shown as a grey box, the dragging event shows an image that is a clone of the block being dragged).

oandregal committed:
Add withDraggable HOC

oandregal committed:
Deprecate Draggable

nosolosw commented on packages/components/src/higher-order/with-draggable/index.js:
```diff
@@ -0,0 +1,154 @@
+const cloneHeightTransformationBreakpoint = 700;
+const clonePadding = 20;
+
+const withDraggable = createHigherOrderComponent(
```
Note that this code has been ported with no substantial modifications from the `Draggable` component.

nosolosw commented on packages/editor/src/components/block-list/block-draggable.js:
```diff
@@ -4,28 +4,35 @@
+class BlockDraggable extends Component {
+	render() {
+		const { clientId, elementId, index, initDragging, isDragging, layout, rootClientId } = this.props;
+		const className = classnames( 'components-draggable', 'editor-block-list__block-draggable', {
```
Do we need the `components-draggable` class?

oandregal committed:
Convert block-draggable into class

oandregal committed:
Inline Draggable component

oandregal committed:
Use withDraggable

nosolosw commented on packages/editor/src/components/block-list/block-draggable.js:
```diff
@@ -2,30 +2,57 @@
+			layout,
+		};
+		initDragging( elementId, transferData )( event );
+		setTimeout( () => onDragStart( event ) );
```
Note that we need to make sure `onDragStart` is executed after `initDragging`.

This is due to how `Draggable` and Drag and Drop behavior was originally [set up](https://github.com/WordPress/gutenberg/blob/5b81cbc66c5d52f9603f77a60ec7864150173dd0/packages/components/src/draggable/index.js#L130). The `isDragging` prop takes its value from `state.dragging` which is [changed in the `onDragStart` and `onDragEnd` methods](https://github.com/WordPress/gutenberg/blob/5b81cbc66c5d52f9603f77a60ec7864150173dd0/packages/editor/src/components/block-list/block.js#L342) of the block. Upon `isDragging` being true, [the visibility of any block's children is changed to `hidden`](https://github.com/WordPress/gutenberg/blob/5b81cbc66c5d52f9603f77a60ec7864150173dd0/packages/editor/src/components/block-list/style.scss#L95), making them un-reactive to events (so the `initDragging` handler ends up not being set up).

youknowriad commented:
Great work and I can see how decoupling the rendering of the DragHandle is a good idea. Though, there's something bothering me about the `API` of this higher order component. I don't like that the HoC provide us with a `initDragging` component where it's the responsibility of the inner component to call `onDragStart` but it automatically handles `onDragEnd` somehow. 

What I propose is to have this:

```js
const DragHandle = ( { onDragStart, onDragEnd } ) => {
  return (
    <button { ...{ onDragStart, onDragEnd } }>Drag Handle</button>
  );
}

export withDraggable(( props ) => { 
   transfertData: props.data,
   clonedElement: props.clonedElement
} )
```

But if you take a look at this (above), it's the exact API as the `Draggable` component we have right now, only difference is that the drag handle is rendered by the consumer and not the Draggable component itself which mean we can do this:

```js
<Draggable transfertData={ data } clonedElement={ element }>
  { ( { onDragStart, onDragEnd } ) => (
    <button { ...{ onDragStart, onDragEnd } }>Drag Handle</button>
  ) }
</Draggable>
```

And this "backwards compatible" because we can check that if `children` is function we apply the behavior proposed here, if not we just keep rendering the handle in `Draggable`. At the same time we can add a "deprecated" call when `children` is not a function.

Thoughts?

oandregal committed:
Pass props to wrapped component

and let React bind the class methods as well.

oandregal committed:
withDraggable is responsible to call the props passed to the wrapped component

oandregal committed:
BlockDraggable: adapt to new API

oandregal committed:
Fix dragend event handler

oandregal committed:
withDraggable's event handlers take control

To avoid withDraggable's event handlers to be shadowed by the children props
we need to make sure they are passed at the end.

oandregal committed:
Make setTimeout property available within withDraggable

oandregal committed:
Do not pass dragover to children

It seems to be more performant.

API-wise it makes sense that withDraggable takes care of calling
children's dragstart/dragend event handlers as it needs them
to be executed after its own. The reason for this is that the
children's event handlers could modify the DOM element
withDraggable clones, messing up with its behavior.
For example, if the children's event handler change the visibility of
the element that takes the event, in turn the element won't be able
to process any further events.

We don't have the same restrictions with the dragover event.
At that point, it can work independently from the children's.

nosolosw commented:
@youknowriad I like that! It also solves some concerns [I had](https://github.com/WordPress/gutenberg/pull/9311#discussion_r213018651) regarding the API leaking some important details. See 6e4fd6f and 1185662.

Implemented and pushed that API change. I can modify docs later, once we agreed to the API.

youknowriad commented:
I think we can reduce the changes required to implement that if we just add the "children as function" support to the current `Draggable` component.

nosolosw commented:
@youknowriad Commented at the same time :) What do you think of https://github.com/WordPress/gutenberg/pull/9311#issuecomment-417031643 ?

oandregal committed:
Make Draggable ready to process children functions

oandregal committed:
Remove any traces of withDraggable HOC

oandregal committed:
Move deprecation to correct version

oandregal committed:
BlockDraggable: convert it back to a functional component

oandregal committed:
Update docs

oandregal committed:
Update docs

nosolosw commented:
@youknowriad Pushed the necessary changes to use `Draggable` and remove any traces of `withDraggable`.

I took a step back and considered the API with fresh eyes. I think now the data flow is clearer: `Draggable` takes the `onDragStart` / `onDragEnd` props passed to the component (if any) and injects `onDraggableStart` / `onDraggableEnd` to the drag handler. The use of HOCs is so pervasive in Gutenberg that at some point I got used to the indirection they introduce. Your rationale has been very helpful, thanks!

nosolosw commented on docs/reference/deprecated.md:
```diff
@@ -14,6 +14,7 @@ Gutenberg's deprecation policy is intended to support backwards-compatibility fo
 
  - `wp.components.withContext` has been removed. Please use `wp.element.createContext` instead. See: https://reactjs.org/docs/context.html.
  - `wp.coreBlocks.registerCoreBlocks` has been removed. Please use `wp.blockLibrary.registerCoreBlocks` instead.
+- `wp.components.Draggable` as a DOM node drag handler has been deprecated. Please, use `wp.components.Draggable` as a wrap component for your DOM node drag handler.
```
Is there a better way to explain this change?

gziolo replied:
This deprecation will be removed in 4.0 release if it is planned for 3.8 release.

nosolosw replied:
Do we need to include the version that will remove the behavior here? I haven't seen it anywhere else.

nosolosw replied:
Just to make sure: the deprecation is added to the version it was deprecated, right? Or should we move this to the `4.0` section?

gziolo replied:
> The current deprecations are listed below and are grouped by _the version at which they will be removed completely_.

It should be under 4.0 😄

nosolosw replied:
:writing_hand:  32006d3

nosolosw commented on packages/components/src/draggable/index.js:
```diff
@@ -148,6 +149,18 @@ class Draggable extends Component {
+			} );
+		}
+
+		deprecated( 'wp.components.Draggable as a DOM node drag handle', {
```
Same here: is there a better way to explain the change?

youknowriad replied:
Actually, I'm wondering if we should really deprecate this or just consider it as a fallback in case there's no `children` as function prop provided

nosolosw replied:
Unless there is a specific use case where that'd be useful, I'd rather deprecate it. Less code to maintain and understand.

youknowriad commented on packages/editor/src/components/block-list/block-draggable.js:
```diff
@@ -8,24 +8,31 @@ import classnames from 'classnames';
+			onDragEnd={ onDragEnd }
+		>
+			{
+				( { onDraggableStart, onDraggableEnd } ) => (
```
Minor: Maybe we should also pass the `draggable` prop here. That way we can just do `{ ...props }` in the inner div. The idea is: `The Draggable component provides all the necessary props to make dragging work`.

youknowriad replied:
I guess my comment assumes we rename `onDraggableStart` and  `onDraggableEnd` into `onDragStart` and `onDragEnd`

nosolosw replied:
That way the `Draggable`'s `onDragStart`/`onDragEnd` properties would shadow the `BlockDraggable`'s `onDragStart` / `onDragEnd` props. I know we can avoid that by using the `...props` trick but it feels more obscure to me, like a smell of something that's not right.

I like the current proposal better because it has a clearer "API contract": `Draggable` provides event handlers for you to connect in a specific DOM node. That's it.

youknowriad replied:
Fair enough. Noting that it's a common pattern though in the React world (see https://github.com/paypal/downshift)

nosolosw commented on packages/editor/src/components/block-list/block.js:
```diff
@@ -483,14 +483,17 @@ export class BlockListBlock extends Component {
+							rootClientId,
+							clientId,
+							layout,
+						} }
```
This is something that has leaked my next step :)

This PR is the first step in refactoring DnD. The idea is to get rid of the `BlockDraggable` component (an invisible drag handler) in favor of a visible drag affordance. Originally, the proposal was to use the hover label https://github.com/WordPress/gutenberg/issues/7114 but that had some problems and now the plan is to include it within the mover component. That would be my next PR after this is merged.

Unless there is a strong opposition I'd leave it as it is, and we can discuss improvements to that in the next PR.

youknowriad replied:
Actually, I think we should leave the `BlockDraggable` even if it becomes UI-less similarily to the `Draggable` component. It would inject the required `transfertData` prop to the `Draggable` component. That way we can easily have multiple `BlockDraggable` in different places. (Just pass the `uid`)

nosolosw replied:
I like that abstraction. Besides what you mention it also builds up nicely: we could have an `ImageDraggable` component whose only responsibility would be to take care of dragging images around, and so on.

At the moment, I've reverted that change in 8b77faa We can dig deeper into simplifying the input data for `BlockDraggable` in the next step when it becomes UI-less.

oandregal committed:
Clarify BlockDraggable abstraction

It is the component that build the necessary transfer data
to drag a block around.

youknowriad commented on packages/components/src/draggable/README.md:
```diff
@@ -49,10 +51,54 @@ const MyDraggable = () => (
+							onDragStart={ onDraggableStart }
+							onDragEnd={ onDraggableEnd }
+							draggable
+							/>
```
Minor: We should probably indent one tab less here

nosolosw replied:
bbc83ac

youknowriad commented on packages/components/src/draggable/README.md:
```diff
@@ -49,10 +51,54 @@ const MyDraggable = () => (
+							onDragStart={ onDraggableStart }
+							onDragEnd={ onDraggableEnd }
+							draggable
+							/>
```
Same

nosolosw replied:
bbc83ac

youknowriad commented on packages/components/src/draggable/index.js:
```diff
@@ -148,6 +149,18 @@ class Draggable extends Component {
+		}
+
+		deprecated( 'wp.components.Draggable as a DOM node drag handle', {
+			version: 3.8,
```
This should be `3.10` because we need to only wait two releases with the deprecation message and I think we should wait until the `3.7` release to merge this PR.

youknowriad replied:
We should also add `plugin: 'Gutenberg',`

nosolosw replied:
d6e0152

youknowriad commented:
Since this touches packages, we're requiring that we update the `CHANGELOG.md` of each package on PRs :). (This is a recent addition to the workflow).

I left some minor cosmetic comments, This PR is looking and let's merge it after the 3.7 release.

oandregal committed:
Update docs

oandregal committed:
Update deprecation

oandregal committed:
Add note in CHANGELOG

youknowriad reviewed (approved):
LGTM 👍

gziolo commented on packages/components/src/draggable/index.js:
```diff
@@ -148,6 +149,19 @@ class Draggable extends Component {
+		}
+
+		deprecated( 'wp.components.Draggable as a DOM node drag handle', {
+			version: 3.10,
```
`4.0` comes after `3.9` in WordPress world :)

youknowriad replied:
oups forgot about that :)

nosolosw replied:
Fixed in 53b10d3ffec11b4d6eacc57f98639eab86852ce4

oandregal committed:
Update version that will depracate the behavior

oandregal committed:
Add deprecation within 4.0 section

dsifford commented on packages/components/src/draggable/README.md:
```diff
@@ -49,10 +51,54 @@ const MyDraggable = () => (
+							icon="move"
+							onDragStart={ onDraggableStart }
+							onDragEnd={ onDraggableEnd }
+							draggable
```
FYI @nosolosw: the `draggable` attribute is enumerated, so shorthanding it like you have here is actually not valid.

See here: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/draggable

cc @youknowriad

youknowriad replied:
Isn't this something already taken care of by React? we should fix if not :)

nosolosw replied:
@dsifford you're right in that `draggable` is enumerated. We could probably the example (and [code](https://github.com/WordPress/gutenberg/blob/7c1811268d47181b5a4d77497706ddeb948f6cd1/packages/block-editor/src/components/block-mover/drag-handle.js#L32)) to remain as true to the HTML standard as possible. However, note that React does the [conversion](https://github.com/facebook/react/blob/d38cfd452f014167d223c73091d4a148a9f893a5/packages/react-dom/src/shared/DOMProperty.js#L252) for us.
